### PR TITLE
SWATCH-2521: Remove use of subscription_product_ids table

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: SWATCH-2521

## Description
At this time, subscription contains the required fields from subscription_product_ids. Data can be obtained from the subscriptions table by:

- sku->sku_product_tag equivalent to subscription_product_ids>product_id

Note that I haven't dropped the table because it would be too dangerous (we couldn't revert this change if we spot something). So, I will create a follow-up JIRA task for this if the changes look good.

## Testing
Only regression testing.

### CODE REVIEW

**Review by kartikshahc — COMMENTED** [review]
I have left few comments. Once those are answered I can go ahead and approve or check for changes.

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionEntity.java [hunk 1]
```diff
@@ -190,33 +169,11 @@ public int hashCode() {
...
-    spi.setSubscription(this);
-    subscriptionProductIds.add(spi);
-  }
-
   public void addSubscriptionMeasurement(SubscriptionMeasurementEntity sm) {
     sm.setSubscription(this);
     subscriptionMeasurements.add(sm);
   }
 
-  public void addSubscriptionMeasurements(Collection<SubscriptionMeasurementEntity> measurements) {
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionEntity.java:213]: This method was not used, so I took advantage of removing dead code.

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionEntity.java [hunk 2]
```diff
@@ -190,33 +169,11 @@ public int hashCode() {
     }
   }
 
-  public void endSubscription() {
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/SubscriptionEntity.java:193]: This method was not used, so I took advantage of removing dead code.

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java [hunk 3]
```diff
@@ -51,13 +50,16 @@ public class MeasurementMetricIdTransformer {
    * @param subscription subscription w/ measurements in the cloud-provider units
    */
   public void translateContractMetricIdsToSubscriptionMetricIds(SubscriptionEntity subscription) {
+    if (subscription.getOffering() == null) {
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java:53]: I prefer the natural null checking :)

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java [hunk 4]
```diff
@@ -51,13 +50,16 @@ public class MeasurementMetricIdTransformer {
...
+    }
+
     try {
       if (subscription.getBillingProvider() == BillingProvider.AWS) {
-        var tag =
-            subscription.getSubscriptionProductIds().stream()
-                .findFirst()
-                .map(SubscriptionProductIdEntity::getProductId)
-                .orElseThrow();
+        var tag = subscription.getOffering().getProductTags().stream().findFirst().orElseThrow();
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java:62]: If you have the another card handy, I'm happy to add a comment here.

File: src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java [hunk 5]
```diff
@@ -205,7 +205,7 @@ protected List<CapacitySnapshotByMetricId> getCapacitiesByMetricId(
     var dbReportCriteria =
         DbReportCriteria.builder()
             .orgId(orgId)
-            .productId(productId.toString())
+            .productTag(productId.toString())
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java:208]: I think we're spoken about mixing the product IDs, product tags and eng ids concepts in many places in the code :(
Actually, from my point of view the ProductId enum should be renamed to either simply Product or ProductTag (or the other way around). 
Anyway, well spotted about the `productId.toString`, will rename it to simply `product`.

**Review by kartikshahc — APPROVED** [review]
Looks good. Once you create a card go ahead and merge this. 
I believe QE should already have automation in place to test this code path so, QE might not be required.

File: src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sgitario) [reply src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java]: I also created https://issues.redhat.com/browse/SWATCH-2663 to address this issue.

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java:62]: I've just seen that in the method resolveConflictingMetrics, we're looping over all the product tags, so I think it makes sense to do the same here, so don't need to create a JIRA ticket for this.

**Comment by kartikshahc** [comment]
LGTM

**Comment by Sgitario** [comment]
Confirmed that the failure [tests.swatch-contracts.test_contracts_api.test_contracts_api_get_contract_filter](https://ci.ext.devshift.net/job/RedHatInsights-rhsm-subscriptions-pr-check/7249/testReport/junit/tests.swatch-contracts/test_contracts_api/test_contracts_api_get_contract_filter/) is also failing for other PRs. Merging.

**Pull request merged**